### ISSUE TITLE
Some drun tests do not need to be drun tests (misleading directory name?)

### ISSUE THREAD

**Issue by osa1** [dfinity/motoko#2687]
Tests below are in run-drun/ directory, but they're not run with drun or ic-ref so they either not need to be in drun/ or the test configuration is buggy and they're incorrectly skipped.

- actor-creation
- actor-import
- aritybug
- chat
- control
- create-then-trap
- general-type-components
- inter-query
- issue-1938-b
- issue-1938-c
- issue-1938-d
- issue-1938
- issue-2158
- issue-2319
- pass-references
- reference-params
- scope-example-func-implicit
- scope-example-func
- scope-example-implicit
- scope-example
- self-shadow
- show-in-actor
- test_oneway
- unavailable-constructor
- unsupported-more
- unsupported

Some of them have skip lines for drun and ic-ref. Others skip compilation entirely. Some of them are only type checked even though they don't have any `//SKIP` directives (e.g. `inter-query`, these seem buggy). Some of these are also interpreted but the interpreter can't really interpret canister tests so they don't do anything other than type checking.

I'm guessing over time the directory evolved into something for all canister tests (some of them only type checked, some also run). In that case perhaps we should rename it into "canisters" or something like that.

It would be good to understand why tests like `inter-query` are skipped even though they don't have any `//SKIP` directives.

**Comment by crusso** [comment]
Are you sure they are skipped or just don't do any printing? I think Joachim set the tests up to avoid recording empty output files (and really wanted most tests to just assert success rather than producing output to be diffed).

If you examine the `ok` dir, you'll see that some tests do have output for run, run-ir, and run-low (for example).

**Comment by crusso** [comment]
inter-query is skipped because it doesn't pass the type checker (it uses nested queries which are currently rejected but were on the roadmap for the ic).

**Comment by osa1** [comment]
> Are you sure they are skipped or just don't do any printing?

I think they're skipped. When we run a test using IR interpreter we print `[run-ir]`. If it's not printed then that step is skipped as far as I can see. Relevant code: https://github.com/dfinity/motoko/blob/19889cf6dc4866bce1df93254e3e56fad197c48a/test/run.sh#L99-L107

If I'm right then in the tests listed above we skip drun and ic-ref steps. Example:

```
[nix-shell:~/dfinity/motoko_2/test/run-drun]$ ../run.sh -d chat.mo
chat: [tc] [run] [run-ir] [run-low]
All tests passed.
```

No `drun-run` or `ic-ref-run` here.

So the question is why are these tests in `run-drun` if they're not run in drun.

> inter-query is skipped because it doesn't pass the type checker (it uses nested queries which are currently rejected but were on the roadmap for the ic).

Hm, this is a bit suprising.. So how it works is, if type checking fails we don't run any of the other steps, but if the type checker output and moc exit code is as expected (as specified in "ok" files) then that's OK (test passes). I guess that's fine, I should update my OCaml implementation. (#2684)

I still think at least the name "run-drun" is confusing because it suggests that the tests under that directory should be run under drun. If you look at the Makefile under that directory it runs the tests with `./run.sh -d *.mo *.drun`. The `-d` says "run under drun", but many of these tests (those listed in my original message) cannot be run with drun. I just find it confusing..

I think it makes sense to organize tests based on one of these two:

- Language features: if a test was added primarily to exercise e.g. arrays then it would be in a directory like "array". If we do that then perhaps this directory is fine. All of these tests use actors somehow. The name could perhaps be "actor" or something like that.

- Compiler features. This means if a test was added primarily to exercise a type checking feature or code then it would be in a directory like "typecheck". It's fine if we still run the test with the interpreters, maybe with drun etc.

Currently what we have is based on how we run the tests (each directory has its own Makefile that passes different parameters to run.sh or have their own testing implementation), but there are cases where we call "run under drun" (run.sh -d) but test isn't run under drun which is what confuses me. It looks like a bug that I call `run.sh -d` (so the intention is running under drun) but drun step is skipped.

**Comment by crusso** [comment]
You are correct that the tests have evolved and the name might be misleading.
 
Some of the test are in here because the type system actually depends on the target to some extent and programs that aren't reject by a vanilla --check or the interpreters are actually rejected because of the target.  Either because the platform doesn't yet or never will support them.

Frankly, I'd prefer it if the language we supported was just what is supported by the platform, which would avoid a lot of this confusion.

So maybe the right way to name these directories are actually by the targeted platform:
run -> wasi
drun -> ic
check -> interpreter

Is it hard to maintain the existing semantics (such as it is) in your ocaml port?

**Comment by nomeata** [comment]
> I'm guessing over time the directory evolved into something for all canister tests (some of them only type checked, some also run). In that case perhaps we should rename it into "canisters" or something like that.

Yes, that it is. `run-drun` simply means “Use the IC system api” (which even affects type-checking) instead of `run/`, which means “use the wasi api”. Feel free to rename if you find it helpful.
